### ISSUE TITLE
Decisions 06: number, point and box through a constrained decode

### ISSUE THREAD

**Issue by gpillon** [gpillon/ignis#242]
The primitive Jev does not have: a number read digit by digit, with `point` and `box` over it. Needs a per-lane permitted token set in the leaf's sampling ABI — the host-side alternative costs six synchronizations and 607 KB per number and is rejected in ADR 0034. The only slice that generates tokens or touches the kernel.

**Spec:** `.scratch/decide/specs/06-number-point-box.md`
**Blocked by:** #238. **Part of:** #236

**Comment by gpillon** [comment]
**Partial — the kernel half is done and proven on the card; the serving half is not.** Two commits on `decide-236`: `0d5c442`, `ab8c49e`. Workspace 1543 green, `cargo check --workspace --features cuda --tests` clean, kernel rebuilt.

## Done

**The permitted token set is in the sampling ABI**, as ADR 0034 requires, and no logits cross for it (acceptance 2 holds by construction on this path). `ignis_sampling_params` appends `permitted_count`/`permitted_ids`; `ignis_decode_options` and `ignis_prefill_options` each append one float out-param for the drawn token's share of its own set. ADR 0016's field append and size bump; a count of 0 is bit-for-bit what the ABI did before.

The constraint is applied to the logits the **vendored** sampler is about to read — `ignis_permit_mask` drives every column outside the set to `-1e30` between the traversal and `ninfer::ops::sample` — so ADR 0010's op is untouched and the set *composes*: greedy takes the set's argmax, temperature draws from the set, top-k/top-p cut it further, penalties still read the sequence's counts.

**A constrained round takes the eager path**, whatever decode graph is ready. Adding a node to a graph captured at load — for a path that runs six rounds per number and never in a throughput workload — would put new failure modes on every decode in the process to save a graph launch on almost none. Cost is the submission, measured at 2.2% of a round.

**The design error the GPU test caught.** `ignis_program_decode` returns the successor the *previous* call made ready, so a run that constrained only its rounds commits the prompt's own free successor as the first token of its forced text. The first version of the test asked for a digit and got the prompt's free answer back. So the **prefill** honours a set too: a K-token run is one constrained prefill plus K-1 constrained rounds. That is now in the ABI docs and the test's header, because it is the kind of thing the next person would rediscover the hard way.

Measured (`crates/core/tests/permitted_decode_gpu.rs`):

```
greedy 25 of [18..27], p 0.18234232 (host 0.182342)
eight hot draws committed {18, 20, 21, 23, 24, 25, 26}
```

The set deliberately excludes the prompt's free answer (17), so a mask that did nothing fails rather than passes; the device's probability matches the host's restricted softmax to six decimals; temperature 2.0 spreads eight draws over seven of ten ids and never leaves the set; a lane beside a constrained one is untouched; a set past the 32-id cap and an id past the vocabulary are refused, not truncated.

**Acceptance 1 is met** (greedy and temperature, on the card). **Acceptance 2 is met.**

## Not done

Everything above the `step::` seam:

- the `Compute` seam and `DecodeJob` carrying a per-step set, and `MockCompute` honouring it (this is what makes acceptances 4, 5 and 6 CPU-testable at all);
- the scheduler's **program**: a request carrying a schedule of sets, the cursor, termination when the schedule is exhausted, and the per-step `(token, probability)` trace reaching the handler;
- the wire types `number` / `point` / `box`, `digits` 1..=6 with a 422 outside it, the `uncertainty = Σ(1−p_k)·10^place` with its per-digit trace, and pixels from the submitted image's own dimensions;
- acceptance 3 (a `point` over the committed fixture, matching `classify_pointing_gpu.rs` end to end through the server) and acceptance 5.

Also unsettled and worth your call: **does a `number`/`point`/`box` count in #241's `ignis_decisions_total`?** It is not a readout and has no answer mass, so either it gets its own `type` label values with no mass observation, or the family stays readout-only and this primitive is counted elsewhere. I did not want to widen a contract row I wrote yesterday without you.

Leaving the issue open. The two commits are self-contained: the ABI is additive, nothing above it calls the new entry points yet, and every existing path is unchanged.

**Comment by gpillon** [comment]
## Done — six commits on `decide-236`

`0d5c442` kernel ABI + mask · `ab8c49e` a constrained run starts at the
prefill · `489a309` the `Compute` seam · `3aa36fb` the scheduler's schedule ·
`b10d4dc` the wire · `fcc64d5` the acceptance on the card.

Workspace **1576 passed / 0 failed**; `cargo check --workspace --features cuda
--tests` clean; the two GPU tests pass on the card.

### The six acceptances

| # | Spec | Evidence |
|---|---|---|
| 1 | a constrained decode returns only tokens from its set, greedy and temperature | `crates/core/tests/permitted_decode_gpu.rs` (greedy takes the set's argmax over a set that excludes the prompt's own free answer, so an inert mask fails; 8 temperature draws over 7 of 10 ids) |
| 2 | no logits cross the `Compute` seam | by construction, checked at the types: in, `PrefillJob::permitted` / `DecodeJob::permitted` are `Option<PermittedSet>`; out, `DecodeOutcome::probabilities` is one `f32` per **emitted token**. `PrefillOutcome` is unchanged and `RuntimeCompute::prefill_step` still allocates a vocabulary buffer only for a job that asked for a `readout`. The restricted softmax is computed on the device over at most 32 logits |
| 3 | a `point` lands inside the button on all three scenes, end to end | `crates/server/tests/decide_point_gpu.rs` — and on a `dflash2-7` drafter load too, pixel for pixel |
| 4 | `uncertainty` matches the trace's place-weighted sum | asserted per axis per scene in the GPU test, and re-derived independently from the response's own digits in `decide_http.rs::sigma_of` |
| 5 | a non-square image returns pixels consistent with its own dimensions | `decide_http.rs::a_point_on_a_non_square_image_is_in_that_images_pixels`, 640×360 through `MockCompute`; `PreparedMedia` now carries the **submitted** size, before `smart_resize` |
| 6 | `digits` outside 1..=6 is a 422 | `decide_wire.rs::digits_outside_the_served_range_is_refused` (`digits_out_of_range`, which `Refusal` renders as 422 like every other refusal on this endpoint) |

The served run, against the fixture manifest:

```
[default]   large  (3149,3424) err  84 px (2.1%)  inside
            medium (1279, 988) err  13 px (0.3%)  inside
            small  (3657, 611) err  34 px (0.8%)  inside
[dflash2-7] identical
```

### What the design turned on

- **A constrained run starts at the prefill.** A decode round returns the
  successor the *previous* call made ready, so a run that constrained only
  its rounds commits the prompt's own free successor inside its forced text.
  A K-token run is one constrained prefill and K rounds. The seam states that
  lag rather than hiding it, and `MockCompute` models it on purpose — a mock
  that hid it would let a scheduler with exactly that bug pass every CPU test.
- **`max_tokens` is dropped at `submit`**, not ignored where it is read: the
  backend enforces it before the leaf runs and cannot tell a run's last round
  from an ordinary one.
- **A round with any constrained lane takes the plain path for the whole
  batch.** A drafter load runs every round as a verify round, and the leaf
  refuses a constrained lane there. That is why acceptance 3 is also run on a
  drafter load — otherwise the routing is an untested comment.
- **The forced prefix rides in the prompt**, so `{"x":` costs prefill rather
  than a round each; every later literal is a permitted set of one, drawn at
  probability 1 by arithmetic, and adds nothing to the uncertainty. For a
  multimodal prompt the MRoPE positions have to grow with it
  (`Multimodal::append_text`) or the leaf refuses the chunk — the measured
  test never met this, because it forced its literal through a separate text
  prefill call.

### Metrics — the question from the earlier comment

The owner answered **yes** (told to me directly rather than on the thread, so
recording it here): `number`/`point`/`box` count in `ignis_decisions_total` —
three more `type` values — and observe **no** answer mass, because they have
none: their answer is a run of sampled tokens, not a restricted softmax over
one position. The histogram is therefore readout-only and its `_count` sits
below the counter's sum by exactly the number of constrained decodes served.
ADR 0017 states that in words rather than leaving it to be inferred from a
graph that does not add up.

### Carried forward, stated rather than hidden

- **`number` and `box` prompts are unmeasured.** Only `point` at three digits
  is; `box` is that text extended by the obvious reading of it, and `number`
  declares a range that is the model's own only by coincidence — whether it
  obeys a *declared* range is an open follow-up in the finding. Each says so
  where it is defined.
- **The served run and the direct-kernel run are not the same reading.**
  `small`'s y reads 149 where `classify_pointing_gpu.rs` read 165: a changed
  tens digit, 66 px, on a digit the direct run read at p=0.665. It stays
  inside a 90 px button, by less than the drift. Two candidates and this run
  separates neither — hq-e8-2b vs the test's BF16, and the two different
  rotations of the forced prefix. Filed as a follow-up in the finding (rerun
  on a BF16 load), not attributed.
- **`uncertainty` is self-declared and is not a bound.** On the measured
  sample it covers the true error on four of six axes; the derivation and the
  two misses are now in the finding, and the caveat is on the wire type.
- **ADR 0034 says no logits cross the seam, and they do not — but it does not
  mention the one-round lag** that carrying the draw's probability across the
  seam requires. A one-line amendment or a follow-up; owner's call, and I did
  not want to edit an ADR's claim on my own a second time.
- `to_pixels` maps the top of the range to `width`, one past the last
  addressable column. Same arithmetic the measured test used, so the two
  agree; stated at the function.

Review: two-axis, both agents read-only. Everything actionable is fixed in
this range — the glossary collision (`Program` was already ADR 0009's
leaf-side layer, so the code now uses `CONTEXT.md`'s own **Constrained
decode** vocabulary, and **Permitted set**, **Schedule** and **Draw** are
glossary entries), an inverted assertion in the non-square test that only
passed because its branch never ran, and two per-request allocations plus a
per-lane mutex the seam change had added to the decode path.
